Add initial support for STM32L0 HAL and STM32L072CZ SOC

ekarlso opened the pull request.

ekarlso commented:
cc @galak @ydamigos

ekarlso commented:
I have a example branch that uses this is in a board but where I get an error when doing a build with it:
```In file included from /home/ekarlso/software/zephyr/arch/arm/soc/st_stm32/stm32l0/soc.h:31:0,
                 from /home/ekarlso/software/zephyr/include/arch/arm/cortex_m/cmsis.h:21,
                 from /home/ekarlso/software/zephyr/arch/arm/include/cortex_m/exc.h:30,
                 from /home/ekarlso/software/zephyr/arch/arm/include/kernel_arch_data.h:53,
                 from /home/ekarlso/software/zephyr/kernel/include/kernel_structs.h:59,
                 from /home/ekarlso/software/zephyr/arch/arm/core/offsets/offsets.c:26:
/home/ekarlso/software/zephyr/ext/hal/st/stm32cube/stm32l0xx/soc/stm32l0xx.h:166:3: error: #error "Please select first the target STM32L0xx device used in your application (in stm32l0xx.h file)"
  #error "Please select first the target STM32L0xx device used in your application (in stm32l0xx.h file)"
```

tagunil commented on arch/arm/soc/st_stm32/stm32l0/Kconfig.series:
```diff
@@ -0,0 +1,20 @@
+config SOC_SERIES_STM32L0X
+	bool "STM32L0x Series MCU"
+	select CPU_CORTEX_M
+	select CPU_CORTEX_M0
```
It's a CPU_CORTEX_M0PLUS.

tagunil commented on arch/arm/soc/st_stm32/stm32l0/soc.c:
```diff
@@ -0,0 +1,95 @@
+
+	/* Update CMSIS SystemCoreClock variable (HCLK) */
+	/* At reset, system core clock is set to 8 MHz from HSI */
+	SystemCoreClock = 8000000;
```
This is incorrect. L0s are clocked from MSI by default, correct value is 2097152 (2.097 MHz)

tagunil commented on arch/arm/soc/st_stm32/stm32l0/soc_irq.h:
```diff
@@ -0,0 +1,59 @@
+#define STM32L0_IRQ_USART3_4_5_6		29
+/* reserved */
+#define STM32L0_IRQ_USB			31
+
```
L0 IRQs are quite different, see Reference Manual.

tagunil replied:
Now that's much better, but there is some tabs/spaces mix between names and numbers

erwango replied:
Agreed with @tagunil, to be fixed

tagunil commented:
I'm very interested in porting STM32L0 as well, but my own efforts were too slow to make it by now. I have some L0 boards to test (NUCLEO-L073RZ and B-L072Z-LRWAN1).

ydamigos reviewed (changes requested):
Please fix your commit messages. You need to add a commit message body and 'Signed-Off-By' line. It is also nice on commit header to show where the changes take place (e.g. `ext: hal: stm32l0xx:`, `arch: arm: soc:`).
For every new file you introduce, add your name or organization in copyright.

erwango commented on arch/arm/soc/st_stm32/stm32l0/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,32 @@
+
+endif # GPIO_STM32
+
+if I2C && (I2C_1 || I2C_2 || I2C_3)
```
Let's focus on GPIO/PINMUX/UART for the initial port.
Other IPs, like I2C, SPI could be added one by one once this initial port is merged.
It will reduce the review/rework load for this PR and enable faster merge.

erwango commented on arch/arm/soc/st_stm32/stm32l0/Kconfig.defconfig.stm32l072cz:
```diff
@@ -0,0 +1,22 @@
+# Kconfig - ST Microelectronics STM32L072XB MCU
+#
+# Copyright (c) 2017 BayLibre, SAS
```
Copyright

erwango commented on arch/arm/soc/st_stm32/stm32l0/Kconfig.defconfig.stm32l072cz:
```diff
@@ -0,0 +1,22 @@
+	string
+	default stm32l072cz
+
+config FLASH_PAGE_SIZE
```
Remove flash. It could be added it in a later PR

ekarlso replied:
Done

erwango commented on arch/arm/soc/st_stm32/stm32l0/Kconfig.soc:
```diff
@@ -0,0 +1,16 @@
+# Kconfig - ST Microelectronics STM32L0 MCU line
+#
+# Copyright (c) 2016 Open-RnD Sp. z o.o.
```
Copyright

erwango commented on arch/arm/soc/st_stm32/stm32l0/Kconfig.soc:
```diff
@@ -0,0 +1,16 @@
+prompt "STM32L0x MCU Selection"
+depends on SOC_SERIES_STM32L0X
+
+config SOC_STM32L072CZ
```
Use  SOC_STM32L072**X**Z and "STM32L072**X**Z"
It will allow to factorize some code when new SoCs will be added

erwango commented on arch/arm/soc/st_stm32/stm32l0/dts.fixup:
```diff
@@ -0,0 +1,37 @@
+#define CONFIG_UART_STM32_PORT_2_NAME		ST_STM32_USART_40004400_LABEL
+#define PORT_2_IRQ				ST_STM32_USART_40004400_IRQ_0
+
+#define CONFIG_I2C_1_BASE_ADDRESS		ST_STM32_I2C_V2_40005400_BASE_ADDRESS
```
Remove I2C, SPI and Flash

erwango commented on arch/arm/soc/st_stm32/stm32l0/flash_registers.h:
```diff
@@ -0,0 +1,51 @@
+/*
```
Flash related, remove

erwango commented on arch/arm/soc/st_stm32/stm32l0/linker.ld:
```diff
@@ -0,0 +1,9 @@
+/* linker.ld - Linker command/script file */
+
+/*
+ * Copyright (c) 2014-2016 Wind River Systems, Inc.
```
Copyright

erwango commented on arch/arm/soc/st_stm32/stm32l0/soc.c:
```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2017 RnDity Sp. z o.o.
```
Copyright

erwango commented on arch/arm/soc/st_stm32/stm32l0/soc.c:
```diff
@@ -0,0 +1,95 @@
+/**
+ * @brief Relocate vector table to SRAM.
+ *
+ * On Cortex-M0 platforms, the Vector Base address cannot be changed.
```
Check if valid on CORTEX-M0+ platform

erwango commented on arch/arm/soc/st_stm32/stm32l0/soc.h:
```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2017 RnDity Sp. z o.o.
```
Copyright

erwango commented on arch/arm/soc/st_stm32/stm32l0/soc.h:
```diff
@@ -0,0 +1,60 @@
+#include <stm32l0xx_ll_system.h>
+#endif /* CONFIG_CLOCK_CONTROL_STM32_CUBE */
+
+#ifdef CONFIG_IWDG_STM32
```
Remove this one

erwango commented on arch/arm/soc/st_stm32/stm32l0/soc.h:
```diff
@@ -0,0 +1,60 @@
+#include <stm32l0xx_ll_iwdg.h>
+#endif
+
+#ifdef CONFIG_I2C_STM32_V2
```
Remove

erwango commented on arch/arm/soc/st_stm32/stm32l0/soc.h:
```diff
@@ -0,0 +1,60 @@
+#include <stm32l0xx_ll_i2c.h>
+#endif
+
+#ifdef CONFIG_SPI_STM32
```
Remove

erwango commented on arch/arm/soc/st_stm32/stm32l0/soc_gpio.c:
```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2016 Open-RnD Sp. z o.o.
```
Copyright

erwango commented on arch/arm/soc/st_stm32/stm32l0/soc.h:
```diff
@@ -0,0 +1,60 @@
+ * @file SoC configuration macros for the STM32L0 family processors.
+ *
+ * Based on reference manual:
+ *   STM32L030x4/x6/x8/xC,
```
Find and replace abuse here

erwango commented on dts/arm/st/mem.h:
```diff
@@ -80,6 +80,9 @@
 #elif defined(CONFIG_SOC_STM32L432XC)
 #define DT_FLASH_SIZE		__SIZE_K(256)
 #define DT_SRAM_SIZE		__SIZE_K(64)
+#elif defined(CONFIG_SOC_STM32L072CZ)
```
Put it before L4XX

erwango commented on arch/arm/soc/st_stm32/stm32l0/soc_registers.h:
```diff
@@ -0,0 +1,15 @@
+/*
+ * Copyright (c) 2016 Open-RnD Sp. z o.o.
```
Copyright

erwango commented on dts/arm/st/stm32l0.dtsi:
```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2017 RnDity Sp. z o.o.
```
Copyright

erwango commented on dts/arm/st/stm32l0.dtsi:
```diff
@@ -0,0 +1,121 @@
+			label = "UART_2";
+		};
+
+		i2c1: i2c@40005400 {
```
Remove i2c, spi

erwango commented on dts/arm/st/stm32l072.dtsi:
```diff
@@ -0,0 +1,7 @@
+/*
+ * Copyright (c) 2017 BayLibre, SAS
```
Copyright

erwango reviewed (changes requested):
Thanks for your work on porting L0, this is an interesting series.

Several things to take care:
*We'll need a board port in this pull request to be able to build/test your changes
*Provide drivers for GPIO, PINMUX, CLOCK CONTOL, UART, this will also be required for basic test port (Hello World)
*Code related to other IPs should be removed to limit the review effort
*Split your commits into logical parts, as you started (HAL, SOC, Clock, GPIO, PINMUX, UART, board)
*Provide a comment message and a Sign-off in each commit
*Provide your copyright in the new files you introduce

ekarlso commented:
@erwango But it is not possible to remove the HAL for the stm32l0 since that breaks building of the example board then?

erwango commented:
@ekarlso : To solve your issue, you should define STM32L072XZ instead of STM32L072CZ  in Kconfig.soc
This is used as $(CONFIG_SOC ) in ext/hal/st/stm32cube/CMakeLists.txt
`zephyr_compile_definitions(`
 `- D$(CONFIG_SOC )`

ekarlso commented:
@erwango it's updated now. Please take a look.

tagunil commented on arch/arm/soc/st_stm32/stm32l0/soc_irq.h:
```diff
@@ -0,0 +1,59 @@
+#define STM32L0_IRQ_USART2			28
+#define STM32F0_IRQ_USART3_4_5_6		29
+/* reserved */
+#define STM32F0_IRQ_USB			31
```
Why F0 here?
There are more differences due to different peripherals in L0 such as LPTIM1. Please look at my version based on RM: https://pastebin.com/52a8Li3G

ekarlso replied:
I have check the manual for it and it seems it's

29 36 settable LPUART1 + AES LPUART1 global interrupt through
EXTI28 + AES global interrupt 0x0000_00B4

So I guess it should be removed?

tagunil replied:
I'm referring to "F0" instead of "L0" in some of your defines, looks like copy-paste error.
And many interrupts are really different from F0, just like in your quote. LPUART, LPTIM etc.

tagunil commented on boards/arm/dragino_lsn50/Kconfig.board:
```diff
@@ -0,0 +1,11 @@
+# Kconfig - STM32F072-EVAL board configuration
```
Board name

tagunil commented on boards/arm/dragino_lsn50/Kconfig.defconfig:
```diff
@@ -0,0 +1,21 @@
+# Kconfig - STM32F072-EVAL board configuration
```
Board name

tagunil commented on boards/arm/dragino_lsn50/support/openocd.cfg:
```diff
@@ -0,0 +1,12 @@
+source [find board/stm32f0discovery.cfg]
```
Is that good for your board?

ekarlso replied:
You are correct, how to do that?

tagunil replied:
It is board specific. Does this board have embedded debugger?

erwango replied:
You can - have a try - with st_nucleo_l1.cfg

tagunil replied:
@ekarlso there is no embedded ST-Link on this board, right? How is it flashed?

ekarlso replied:
Done

erwango reviewed (changes requested):
Thanks for the quick update!

For the efficiency of the review, can you please split into several logical commits (HAL, SOC, GPIO, PINMUX, UART, CLOCK, BOARD)?

Also, you'll need to provide Signed-off-by and minimum comment in each commit, else your pull request won't pass the Continuous Integration tests (Shippable).
Please check other similar pull requests to see how it should looks like.

ekarlso committed:
hal: stm32l0x: Add HAL for the STM32L0x series

This adds the HAL code needed for the SoC in a later patch for
the stm32l0 series.

Origin: ST Microelectronics
License: BSD-3-Clause
URL: http://www.st.com/en/embedded-software/stm32cubel0.html
Commit: 1.10.0
Purpose: HAL layer for stm32l0
Maintained-by: External

Signed-off-by: Endre Karlson <endre.karlson@gmail.com>

ekarlso commented:
@erwango what do you mean by driver ? Would you be able for a quick chat ?

tagunil commented on arch/arm/soc/st_stm32/stm32l0/soc_gpio.c:
```diff
@@ -0,0 +1,163 @@
+	u32_t lckr;
+	u32_t afr[2];
+	u32_t brr;
+	u32_t ascr; /* Only present on STM32L0x1, STM32L0x5, STM32L0x6 */
```
There is no STM32L0x5/6, and, AFAIK, there is no such register in STM32L0.

ekarlso replied:
I removed it.

tagunil commented on arch/arm/soc/st_stm32/stm32l0/soc_gpio.c:
```diff
@@ -0,0 +1,163 @@
+
+int stm32_gpio_enable_int(int port, int pin)
+{
+	struct stm32l0x_syscfg *syscfg = (struct stm32l0x_syscfg *)SYSCFG_BASE;
```
Where is stm32l0x_syscfg defined?

tagunil replied:
Here is my version of stm32l0x_syscfg: https://pastebin.com/r3Vyed9b

ekarlso replied:
Ok, thanks I will continue working on it on Saturday.

tagunil commented on arch/arm/soc/st_stm32/stm32l0/soc.c:
```diff
@@ -0,0 +1,96 @@
+/**
+ * @brief Relocate vector table to SRAM.
+ *
+ * On Cortex-M0+ platforms, the Vector Base address cannot be changed.
```
AFAIK there is VTOR on Cortex-M0+, so all this relocation stuff may be unneeded on L0.

ekarlso replied:
Are you sure it can be removed?

tagunil replied:
It can be removed eventually, but that may require some changes in arch/arm/core/cortex_m/prep_c.c and maybe somethere else. I think @b0661 and @superna9999 can help with that.

ekarlso committed:
arm: stm32l0: Add support for stm32l0 soc

Add necessary kconfig and minimal device tree support for
the stm32l072xz variants.

Signed-off-by: Endre Karlson <endre.karlson@gmail.com>

ekarlso committed:
dts: arm: provider support for the stm32l0

Add necessary DTS files for stm32l0x and stm32l0x72

Signed-off-by: Endre Karlson <endre.karlson@gmail.com>

tagunil commented on arch/arm/soc/st_stm32/stm32l0/gpio_registers.h:
```diff
@@ -0,0 +1,77 @@
+	u32_t brr;
+};
+
+union syscfg_cfgr1 {
```
I think we can just leave syscfg_cfgr1 as u32 at the moment.

tagunil replied:
Let's remove this union completely. Maybe we'll add correct L0 version later, but it is unneeded ATM.

tagunil commented on arch/arm/soc/st_stm32/stm32l0/soc_irq.h:
```diff
@@ -0,0 +1,60 @@
+
+#define STM32L0_IRQ_WWDG			0
+#define STM32L0_IRQ_PVD         1
+/* reserved */
```
There is no more reserved number here.

tagunil commented on arch/arm/soc/st_stm32/stm32l0/soc_irq.h:
```diff
@@ -0,0 +1,60 @@
+#define STM32L0_IRQ_SPI2			26
+#define STM32L0_IRQ_USART1			27
+#define STM32L0_IRQ_USART2			28
+/* reserved */
```
So what about LPUART1_AES_RNG and LCD interrupts? I've listed them in my paste https://pastebin.com/52a8Li3G

ekarlso replied:
I will add them too

tagunil commented on arch/arm/soc/st_stm32/stm32l0/soc_irq.h:
```diff
@@ -0,0 +1,60 @@
+#define STM32L0_IRQ_EXTI2_3			6
+#define STM32L0_IRQ_EXTI4_15			7
+#define STM32L0_IRQ_TSC             8
+/* reserved */
```
And here.

tagunil commented on arch/arm/soc/st_stm32/stm32l0/soc_irq.h:
```diff
@@ -0,0 +1,60 @@
+#define STM32L0_IRQ_ADC			12
+#define STM32L0_IRQ_LPTIM1	13
+#define STM32L0_IRQ_USART_4_5   14
+/* reserved */
```
And here.

tagunil commented on arch/arm/soc/st_stm32/stm32l0/gpio_registers.h:
```diff
@@ -0,0 +1,77 @@
+	} bit;
+};
+
+struct stm32l0x_syscfg {
```
These registers are different in L0. I would replace that with https://pastebin.com/r3Vyed9b

ekarlso replied:
Will fix.

ekarlso replied:
This was fixed.

tagunil commented on arch/arm/soc/st_stm32/stm32l0/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,27 @@
+
+if GPIO_STM32
+
+config GPIO_STM32_PORTD
```
I'm not sure about this ports. L0s have ports A-E and H, and some of them have even less.

tagunil commented on arch/arm/soc/st_stm32/stm32l0/Kconfig.defconfig.stm32l072xz:
```diff
@@ -0,0 +1,19 @@
+
+config SOC
+	string
+	default stm32l072xz
```
Let's try stm32l072xx here - it can help with build issues.

ekarlso committed:
drivers: gpio: Provide GPIO driver for stm32l0x

Add necessary GPIO defines for the stm32l0x series soc

Signed-off-by: Endre Karlson <endre.karlson@gmail.com>

ekarlso committed:
driver: pinmux: Add pinmux driver for stm32l0x

Add driver header files for the stm32l0x pinmux

Signed-off-by: Endre Karlson <endre.karlson@gmail.com>

ekarlso committed:
drivers: clock_control: Provide support for stm32l0x

Add Clock Control support for the stm32l0x based on reference manuals

Signed-off-by: Endre Karlson <endre.karlson@gmail.com>

miyatsu commented:
I noticed that the board you are porting is a LoRa board, do you have plans to porting LoRaMac to zephyr ? I'm trying to do that, but I'm a bit struggling alone and I want to give up...

ekarlso committed:
driver: interrupt_controller: Add support for stm32l0x

Add kconfig and c code for stm32l0x interrupt controller driver

Signed-off-by: Endre Karlson <endre.karlson@gmail.com>

erwango commented on arch/arm/soc/st_stm32/stm32l0/soc_registers.h:
```diff
@@ -0,0 +1,16 @@
+#define _STM32L0X_SOC_REGISTERS_H_
+
+/* include register mapping headers */
+//#include "flash_registers.h"
```
Clean up the commented line

ekarlso replied:
Done

erwango commented on dts/arm/st/stm32l0.dtsi:
```diff
@@ -0,0 +1,88 @@
+				label = "FLASH_STM32";
+				reg = <0x08000000 DT_FLASH_SIZE>;
+
+				write-block-size = <2>;
```
write-block-size = <4>;

ekarlso replied:
Done

erwango commented on dts/arm/st/stm32l0.dtsi:
```diff
@@ -0,0 +1,88 @@
+		usart1: serial@40013800 {
+			compatible = "st,stm32-usart", "st,stm32-uart";
+			reg = <0x40013800 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB1_2 0x00004000>;
```
STM32_CLOCK_BUS_APB2_1

ekarlso replied:
Done

erwango commented on drivers/gpio/gpio_stm32.h:
```diff
@@ -56,6 +56,14 @@
 #define STM32_PERIPH_GPIOJ LL_AHB1_GRP1_PERIPH_GPIOJ
 #define STM32_PERIPH_GPIOK LL_AHB1_GRP1_PERIPH_GPIOK
+#elif CONFIG_SOC_SERIES_STM32L0X
+#define STM32_CLOCK_BUS_GPIO STM32_CLOCK_BUS_AHB1
```
STM32_CLOCK_BUS_IOP

ekarlso replied:
Done

erwango commented on boards/arm/dragino_lsn50/pinmux.c:
```diff
@@ -0,0 +1,34 @@
+/* pin assignments for Dragino LSN50 board */
+static const struct pin_config pinconf[] = {
+#ifdef CONFIG_UART_STM32_PORT_1
+	{STM32_PIN_PD5, STM32L0_PINMUX_FUNC_PD5_USART2_TX},
```
This does not match what you provided in device tree.
No functional impact since pinmux device tree is not yet enabled, but you should make it be coherent.

erwango reviewed (changes requested):
This is progressing well.

Some additional comments inline.

General comments:
-remove former copyrights in the files you add.
-Provide comment in each commit

ydamigos commented on arch/arm/soc/st_stm32/stm32l0/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,27 @@
+
+config GPIO_STM32_PORTB
+	default y
+
```
Ports A, B and C are enabled from common/Kconfig.defconfig.series

tagunil replied:
@erwango, @ydamigos, what is the policy about ports listed here? Should it be the least common denominator for all L0s, or only for supported at the moment?

ydamigos replied:
We do it only for the supported boards. Some SoCs may never be ported to Zephyr.

tagunil replied:
So looks like we should add only port H for STM32L072CZ.

erwango replied:
> We do it only for the supported SoCs. Some SoCs may never be ported to Zephyr.

@ydamigos: While this is indeed what we have done for GPIO flags clean up,  this rule is not quite obvious from simple code review and might be source of error for new comers. Let keep it that way for now, but we might consider reverting to "all series" rule vs current "currently supported SoC" if there are too much errors linked to that during reviews. What do you think?

ydamigos replied:
>Let keep it that way for now, but we might consider reverting to "all series" rule vs current "currently supported SoC" if there are too much errors linked to that during reviews. What do you think?

@erwango I am fine with your proposal.

tagunil commented on arch/arm/soc/st_stm32/stm32l0/gpio_registers.h:
```diff
@@ -0,0 +1,80 @@
+	} bit;
+};
+
+union syscfg__exticr {
```
Should we fix that for other SOCs in a separate PR?

erwango replied:
Yes, it should be done in a separate PR.
We have enough things to review in this one :-) and PRs should be kept consistent

ekarlso replied:
Done.

ekarlso commented on arch/arm/soc/st_stm32/stm32l0/soc_gpio.c:
```diff
@@ -0,0 +1,145 @@
+{
+	volatile struct stm32l0x_syscfg *syscfg =
+		(struct stm32l0x_syscfg *)SYSCFG_BASE;
+	volatile union syscfg__exticr *exticr;
```
Done

ekarlso commented on arch/arm/soc/st_stm32/stm32l0/soc_gpio.c:
```diff
@@ -0,0 +1,145 @@
+		device_get_binding(STM32_CLOCK_CONTROL_NAME);
+
+	struct stm32_pclken pclken = {
+		.bus = STM32_CLOCK_BUS_APB1_2,
```
Done

tagunil replied:
Not fixed yet

tagunil commented on drivers/interrupt_controller/Kconfig.stm32:
```diff
@@ -13,31 +13,31 @@ config EXTI_STM32
 	help
 	 Enable EXTI driver for STM32 line of MCUs
 
+
```
Why that empty line?

ekarlso replied:
removed

tagunil commented on drivers/interrupt_controller/Kconfig.stm32:
```diff
@@ -13,31 +13,31 @@ config EXTI_STM32
 config EXTI_STM32_EXTI1_0_IRQ_PRI
 	int "EXTI1:0 IRQ priority"
 	depends on EXTI_STM32
-	depends on SOC_SERIES_STM32F0X
```
Should be:
depends on SOC_SERIES_STM32F0X || SOC_SERIES_STM32L0X
as 3_2 and 15_4

ekarlso replied:
fixed.

tagunil commented on drivers/interrupt_controller/exti_stm32.c:
```diff
@@ -154,6 +156,23 @@ void stm32_exti_enable(int line)
 		}
 	}
+#elif defined(CONFIG_SOC_SERIES_STM32L0X)
+		if (line >= 4 && line <= 15) {
```
Fix that formatting, please.

tagunil commented on drivers/interrupt_controller/exti_stm32.c:
```diff
@@ -154,6 +156,23 @@ void stm32_exti_enable(int line)
+	} else {
+		/* > 15 are not mapped on an IRQ */
+		/*
+		 * On STM32F0X, this function also support enabling EXTI
```
L0X

tagunil commented on arch/arm/soc/st_stm32/stm32l0/Kconfig.series:
```diff
@@ -0,0 +1,21 @@
+	select SOC_FAMILY_STM32
+	select SYS_POWER_LOW_POWER_STATE_SUPPORTED
+	select HAS_STM32CUBE
+	select CPU_HAS_MPU
```
Not all L0s have MPU. Let's do that on per-SOC basis. L073 has it.

ekarlso replied:
Ok, we can do that in seperate PRs then..

erwango commented on boards/arm/stm32f072_eval/pinmux.c:
```diff
@@ -12,11 +12,15 @@
 #include "pinmux/stm32/pinmux_stm32.h"
 
-/* pin assignments for STM32F072-EVAL board */
+/* pin assignments for Dragino LSN50 board */
```
Code is good but file name is wrong it seems

erwango commented:
@ekarlso : Good progress, but let's solve the compilation issues before continuing the review.
Then you'll need to provide a doc for your board, otherwise it won't be merged. Doc review can take a little while, so you'd better start soon. You can check what has been done for other boards.

ekarlso committed:
boards: arm: Add Dragino LSN50 LoRA node

Add dts, kconfig and other necessary code for LSN50 board from Dragino
to work with Zephyr.

Signed-off-by: Endre Karlson <endre.karlson@gmail.com>

tagunil commented on dts/arm/st/stm32l0-pinctrl.dtsi:
```diff
@@ -0,0 +1,27 @@
+			};
+			usart2_pins_a: usart2@0 {
+				rx_tx {
+					rx = <STM32_PIN_PA3 (STM32_PINMUX_ALT_FUNC_0 | STM32_PUPDR_NO_PULL)>;
```
ALT_FUNC_4 for both TX and RX

tagunil commented on dts/arm/st/stm32l0.dtsi:
```diff
@@ -0,0 +1,88 @@
+
+		cpu@0 {
+			device_type = "cpu";
+			compatible = "arm,cortex-m0";
```
cortex-m0+

erwango commented:
@ekarlso : Seems it compiles now! 
But please move your picture under boards/arm/dragino_lsn50/doc/**img** to allow doc build...

dbkinder commented on boards/arm/dragino_lsn50/doc/dragino_lsn50.rst:
a list (and nested list) needs a blank line before the first item

dbkinder commented on boards/arm/dragino_lsn50/doc/dragino_lsn50.rst:
add a blank line here too

dbkinder commented on boards/arm/dragino_lsn50/doc/dragino_lsn50.rst:
blank line before this list too

dbkinder commented on boards/arm/dragino_lsn50/doc/dragino_lsn50.rst:
blank line here (and for the nested lists below too)

dbkinder commented on boards/arm/dragino_lsn50/doc/dragino_lsn50.rst:
outdent these, and add a blank line before the first list item

dbkinder reviewed (changes requested):
thanks for adding documentation.  some recommended edits for formatting mishaps.

erwango commented:
@ekarlso : Can you fix errors an warning here:
https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/10675/4/tests

dbkinder reviewed (dismissed):
thanks

superna9999 commented on ext/hal/st/stm32cube/stm32l0xx/CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2017 Erwin Rol <erwin@erwinrol.com>
+# Copyright (c) 2018 Endre Karlson <endre.karlson@gmail.com>
```
This is not the correct location of the change, please do it in the previous patch

superna9999 reviewed (dismissed):
A single change needed, LGTM after this fix

superna9999 commented:
Enabling SPI and I2C should be easy here, and flash support aswell. Did you test it ?

erwango commented:
> Enabling SPI and I2C should be easy here, and flash support aswell. Did you test it ?

@superna9999 : Some part of SPI and I2C where available in the initial version of the PR.
I suggested @ekarlso to reduce the scope (focus on the minimum IP set to enable basic tests) of his PR to ease and fasten the review/merge process.
Once it is merged @ekarlso and other interested parts can add further support to this series.

erwango commented on boards/arm/dragino_lsn50/dragino_lsn50.dts:
```diff
@@ -0,0 +1,33 @@
+
+/ {
+	model = "Dragino LSN50 LoRA Sensor Node";
+	compatible = "st,stm32l072xz";
```
compatible = "_vendor_,dragino", "st,stm32l072";
Set __vendor__ based on actual board vendor

erwango commented on boards/arm/dragino_lsn50/dragino_lsn50.dts:
```diff
@@ -0,0 +1,33 @@
+	status = "ok";
+};
+
+&usart2 {
```
Remove usart2 if not enabled in defconfig

erwango commented on dts/arm/st/stm32l0.dtsi:
```diff
@@ -0,0 +1,87 @@
+#include <arm/armv6-m.dtsi>
+#include <st/mem.h>
+#include <dt-bindings/clock/stm32_clock.h>
+#include <dt-bindings/i2c/i2c.h>
```
Remove this line

erwango reviewed (changes requested):
Couple of minor last changes and we're good to go!

erwango commented on arch/arm/soc/st_stm32/stm32l0/Kconfig.series:
```diff
@@ -0,0 +1,19 @@
+	select CPU_HAS_SYSTICK
+	select CLOCK_CONTROL_STM32_CUBE if CLOCK_CONTROL
+	help
+		Enable support for STM32L0 MCU series
```
Use 2 spaces before help message instead of tab

erwango reviewed (changes requested):
Some comments not addressed and one additional comment. So close

galak reviewed (dismissed):
The ext Hal commit message needs more info.  It should look something like commit 81d4342dc0e1234adb294a426298853f8d0d4aeb

erwango reviewed (changes requested):
One last remaining comment is still not addressed in following file
arch/arm/soc/st_stm32/stm32l0/soc_irq.h

Otherwise, I'm fine with this.

ydamigos commented on arch/arm/soc/st_stm32/stm32l0/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,23 @@
+config SOC_SERIES
+	default stm32l0
+
+if GPIO_STM32
```
Why ports D and E are missing?

ekarlso replied:
Because I was told that they where provided by the stm32 config?

erwango replied:
This is true for A B C only.
E and D are needed there indeed

tagunil replied:
Are there ports D and E in all L072s? Or that doesn't matter?

erwango replied:
Sorry I've been too fast, you're right.
Some L0xx support only ABC, so DEH should go in Kconfig.defconfig.stm32l072xz.

erwango commented:
recheck

erwango commented on arch/arm/soc/st_stm32/stm32l0/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,23 @@
+
+if GPIO_STM32
+
+config GPIO_STM32_PORTH
```
To sum up discussions, Move H in Kconfig.defconfig.stm32l072xz and we're good

erwango commented:
Approved, but requires a rebase.
You can alos remove the [WIP]
